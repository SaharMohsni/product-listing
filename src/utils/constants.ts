export const ASC_PRICE = 0;
export const DESC_PRICE = 1;
export const NEW_ADDED = 2;
export const OLD_ADDED = 3;

export const LIMIT_PRODUCTS_BY_Page = 16;

export const sortDataMessages = {
	[ASC_PRICE]: 'Price low to high',
	[DESC_PRICE]: 'Price high to low',
	[NEW_ADDED]: 'New to old',
	[OLD_ADDED]: 'Old to new'
};

export const MIN_PRODUCT_QUANTITY_IN_BASKET = 1;
export const FAKE_DATA = Array.from({ length: 16 }).map(() => {}).map((el, i) => ({
	tags: [ '' ],
	price: i,
	name: 'string',
	description: 'string',
	slug: 'string',
	added: i,
	manufacturer: i,
	itemType: 'string'
}));
export const DELAY_TIME = 3000;
