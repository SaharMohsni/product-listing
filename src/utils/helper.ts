import { ISearchProductsPayload } from '../features/types/products.types';
import { isEmpty } from 'lodash';
export const getURLCleanPath = (pathname: string) => pathname.split('/')[1];

export const serialize = function(obj: any) {
	var str = [];
	for (var p in obj)
		if (obj.hasOwnProperty(p)) {
			str.push(encodeURIComponent(p) + '=' + encodeURIComponent(obj[p]));
		}
	return str.join('&');
};

export const handleNavigationQuery = (params: ISearchProductsPayload) => {
	let { page, sortVariable, sortType } = params;
	let query = params;
	if (!isEmpty(params) && (isEmpty(sortVariable) || isEmpty(sortType))) {
		query = {
			page: page
		};
	} else {
		query = {
			...params
		};
	}

	return `${serialize(query)}`;
};
