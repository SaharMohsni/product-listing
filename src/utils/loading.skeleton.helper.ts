export const fullRowItemSkeleton = (loading:boolean, active = true) => ({
    active,
    title: { rows: 1, width: '100%' },
    paragraph: false,
    loading,
  })
  export const labelSkeleton = (loading:boolean, active = true) => ({
    active,
    title: false,
    paragraph: { rows: 1, width: '20%'  },
    loading,
  })
  export const tagSkeleton = (loading:boolean, active = true) => ({
    active,
    title: false,
    paragraph: { rows: 1 },
    loading,
  })
  export const imageSkeleton = (loading:boolean, active = true ) => ({
    loading,
  active,
  paragraph: false,
  title:false
  })

  export const descriptionSkeleton = (loading:boolean, active = true) => ({
    active,
    title: false,
    paragraph: { rows: 3, width: '100%' },
    loading,
  })
  export const shapeSquareBoxSkeleton = (loading:boolean, active = true) => ({
    active,
    title: false,
    paragraph: false,
    loading,
  })
