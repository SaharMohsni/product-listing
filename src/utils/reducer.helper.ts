import { ISearchProductsPayload } from '../features/types/products.types';

export const handleSearchParams = (localSearchParams: ISearchProductsPayload, params: ISearchProductsPayload) => {
	let { page, limit, sortVariable, sortType } = localSearchParams;
	return {
		...localSearchParams,
		page: params.page ? params.page : page,
		limit: params.limit ? params.limit : limit,
		sortVariable: params.sortVariable ? params.sortVariable : sortVariable,
		sortType: params.sortType ? params.sortType : sortType
	};
};
