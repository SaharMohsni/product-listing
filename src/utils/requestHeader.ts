 
export const requestHeader = (options: object| null ): object => ({
  headers: { 
    // 'Access-Control-Allow-Origin': '*', 
    ...options,
  },
})

 