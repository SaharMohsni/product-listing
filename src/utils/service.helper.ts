import { isEmpty } from 'lodash';
export const addSortToUrl = (variable: string | undefined, sortType: string | undefined) => {
	if (!isEmpty(variable) && !isEmpty(sortType)) {
		return `&_sort=${variable}&_order=${sortType}`;
	} else return '';
};

export const addFilterToUrl = (variable: any | undefined) => {
	let str = [];
	if (variable) {
		for (var el in variable.data) {
			str.push(variable.key + '=' + variable.data[el]);
		}
	}
	return str.join('&');
};

export const addFilterByArrayListToUrl = (variable: any | undefined) => {
	let str = [];
	if (variable) {
		for (var el in variable.data) {
			str.push(variable.key + '_like=' + variable.data[el]);
		}
	}
	return str.join('&');
};

export const addFilterByStringToUrl = (variable: any | undefined) => {
	if (!isEmpty(variable)) {
		return `itemType=${variable}`;
	} else return '';
};
