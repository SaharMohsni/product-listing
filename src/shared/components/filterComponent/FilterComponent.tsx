/**
 *
 * Filter component : reuseable component
 *
 */

import React, { useState } from 'react';
import { Checkbox, Skeleton, Input, Empty } from 'antd';
import { useLocation } from 'react-router-dom';

import './filter-component.css';

import * as skeleton from '../../../utils/loading.skeleton.helper';
import { createCheckedAllDataStructure, formatFilterDataStructure, getCheckboxCurrentValues } from './helper';
import { isEmpty } from 'lodash';
import { convertObjectKey, generateQueryBaseStructure, generateQueryFromPathname } from '../../../utils/helper';

export interface IOwnProps {
	loading: boolean;
	title: string;
	optionsList: { id: number; value: string }[];
	setSearchedValue: (value: string) => void;
	filterKey: string;
	setFilterData: (value: object) => void;
}
const FilterComponent: React.FC<IOwnProps> = ({
	loading,
	title,
	optionsList,
	setSearchedValue,
	filterKey,
	setFilterData
}) => {
	const [ checkAll, setCheckAll ] = useState(false);
	const location = useLocation();
	let query = generateQueryFromPathname(location.pathname);
	let searchQuery = generateQueryBaseStructure(convertObjectKey(query));

	const handleSearch = (data: any) => {
		let filterData = formatFilterDataStructure(filterKey, data);
		setFilterData(filterData);
	};
	const handleChange = (e: any) => {
		handleSearch(e);
	};
	const onCheckAllChange = (e: any) => {
		setCheckAll(e.target.checked);
		let afterCheckAllData = e.target.checked ? createCheckedAllDataStructure(optionsList) : [];
		handleSearch(afterCheckAllData);
	};

	const handleSearchInputChange = (e: any) => {
		setSearchedValue(e.target.value);
	};
	return (
		<div className="filter-component ">
			<Skeleton {...skeleton.labelSkeleton(loading)}>
				<div className="section-block-header">{title}</div>
			</Skeleton>
			<Skeleton avatar={{ shape: 'square' }} {...skeleton.shapeSquareBoxSkeleton(loading)}>
				<div className="filter-component__data section-block">
					<div>
						<div className="filter-component_search-bar-container">
							<Input
								className="filter-component_search-bar-container__search-bar"
								placeholder={`search ${title}`}
								onChange={handleSearchInputChange}
							/>
						</div>
						{isEmpty(optionsList) ? (
							<Empty image={Empty.PRESENTED_IMAGE_SIMPLE} description="No data" />
						) : (
							<div className="filter-component__filter-options-container box global-scroll">
								<Checkbox onChange={onCheckAllChange} checked={checkAll}>
									All
								</Checkbox>
								<Checkbox.Group
									style={{ width: '100%' }}
									onChange={(e) => handleChange(e)}
									className="global-flex-column-h-start-v-start "
									value={getCheckboxCurrentValues(searchQuery, filterKey)}
								>
									{optionsList.map((option: any) => {
										return (
											<Checkbox key={option.id} value={option.id}>
												{option.value}
											</Checkbox>
										);
									})}
								</Checkbox.Group>
							</div>
						)}
					</div>
				</div>
			</Skeleton>
		</div>
	);
};
export default FilterComponent;
