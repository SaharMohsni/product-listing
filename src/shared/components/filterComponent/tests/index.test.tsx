import React from 'react';
import { render } from '../../../../utils/testing';
import FilterComponent, { IOwnProps } from '../FilterComponent';

describe('Render FilterComponent', () => {
	const defaultProps: IOwnProps = {
		setSearchedValue() {
			return;
		},
		loading: true,
		title: '',
		optionsList: [ { id: 0, value: '' } ],
		filterKey: '',
		setFilterData() {
			return;
		}
	};
	const { container } = render(<FilterComponent {...defaultProps} />);
	test('should render the basic template', async () => {
		expect(container).toMatchSnapshot();
	});
});
