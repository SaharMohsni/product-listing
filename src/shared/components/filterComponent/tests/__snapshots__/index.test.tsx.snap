// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateOrUpdateFacility should render the basic template 1`] = `
<div>
  <div
    class="filter-component "
  >
    <div
      class="ant-skeleton ant-skeleton-active"
    >
      <div
        class="ant-skeleton-content"
      >
        <ul
          class="ant-skeleton-paragraph"
        >
          <li
            style="width: 20%;"
          />
        </ul>
      </div>
    </div>
    <div
      class="ant-skeleton ant-skeleton-with-avatar ant-skeleton-active"
    >
      <div
        class="ant-skeleton-header"
      >
        <span
          class="ant-skeleton-avatar ant-skeleton-avatar-lg ant-skeleton-avatar-square"
        />
      </div>
    </div>
  </div>
</div>
`;
